tests/unit/sources/test_base.py

```diff
@@ -0,0 +1,227 @@
+    def test_pull_file_checksum_error(self):
+        self.source.source = "src/my_file"
+        self.source.source_checksum = "md5/12345"
+        Path("src").mkdir()
+        Path("src/my_file").write_text("content")
+        Path("source/dir").mkdir(parents=True)
+
+        with pytest.raises(errors.ChecksumMismatch) as raised:
+            self.source.pull()
+        assert raised.value.expected == "12345"
+        assert raised.value.obtained == "9a0364b9e99bb480dd25e1f0284c8555"
+
+    def test_pull_url(self, requests_mock, new_dir):
+        self.source.source = "http://test.com/some_file"
+        requests_mock.get(self.source.source, text="content")
+        Path("source/dir").mkdir(parents=True)
+
+        self.source.pull()
+
+        assert self.source.provision_dst == "source/dir"
+        assert self.source.provision_clean_target is False
+        assert self.source.provision_keep is False
+        assert self.source.provision_src == "source/dir/some_file"
+
+        downloaded = Path(new_dir, "source", "dir", "some_file")
+        assert downloaded.is_file()
+
+    def test_pull_url_checksum(self, requests_mock, new_dir):
+        self.source.source = "http://test.com/some_file"
+        self.source.source_checksum = "md5/9a0364b9e99bb480dd25e1f0284c8555"
+        requests_mock.get(self.source.source, text="content")
+        Path("source/dir").mkdir(parents=True)
+
+        self.source.pull()
+
+        assert self.source.provision_dst == "source/dir"
+        assert self.source.provision_clean_target is False
+        assert self.source.provision_keep is False
+        assert self.source.provision_src == "source/dir/some_file"
+
+        downloaded = Path(new_dir, "source", "dir", "some_file")
+        assert downloaded.is_file()
+
+        file_cache = cache.FileCache("app")
+        cached = file_cache.get(key=self.source.source_checksum)
+        assert cached is not None
+        assert Path(cached).read_bytes() == b"content"
+
+    def test_pull_url_checksum_cached(self, requests_mock, new_dir):
+        self.source.source = "http://test.com/some_file"
+        self.source.source_checksum = "md5/9a0364b9e99bb480dd25e1f0284c8555"
+        Path("source/dir").mkdir(parents=True)
+        requests_mock.get(self.source.source, text="other_content")
+
+        # pre-cache this file
+        Path("my_file").write_text("content")
+        file_cache = cache.FileCache("app")
+        file_cache.cache(filename="my_file", key=self.source.source_checksum)
+
+        self.source.pull()
+
+        assert self.source.provision_dst == "source/dir"
+        assert self.source.provision_clean_target is False
+        assert self.source.provision_keep is False
+        assert self.source.provision_src == "source/dir/some_file"
+
+        downloaded = Path(new_dir, "source", "dir", "some_file")
+        assert downloaded.is_file()
+        assert downloaded.read_bytes() == b"content"
+
+    def test_file_source_abstract_methods(self):
```

Comment:
This is just testing python semantics?

If the purpose is to verify abstract methods, perhaps just
`assert FileSourceHandler.__abstractmethods__ == {"method1...",}`

Reply:
We just wanted to make very explicit that subclasses not implementing those methods won't be allowed, but yes, just checking the abstract methods should have been enough.